internal/middlewares/service_injector.go

```diff
@@ -0,0 +1,60 @@
+package middlewares
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/Renal37/go-musthave-diploma-tpl/internal/models"
+)
+
+// Тип для ключей контекста, используемых для хранения сервисов.
+type key int
+
+// Константы для ключей различных сервисов, которые будут храниться в контексте.
+const (
+	AuthServiceKey key = iota
+	JwtServiceKey
+	OrderServiceKey
+	AccrualServiceKey
+	BalanceServiceKey
+)
+
+// ServiceInjectorMiddleware - middleware для инъекции сервисов в контекст запроса.
+// Принимает на вход необходимые сервисы и добавляет их в контекст.
+func ServiceInjectorMiddleware(
+	authService models.AuthService,
+	jwtService models.JWTService,
+	orderService models.OrderService,
+	accrualService models.AccrualService,
+	balanceService models.BalanceService,
+) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Добавляем сервисы в контекст запроса
+			ctx := context.WithValue(r.Context(), AuthServiceKey, authService)
+			ctx = context.WithValue(ctx, JwtServiceKey, jwtService)
+			ctx = context.WithValue(ctx, OrderServiceKey, orderService)
+			ctx = context.WithValue(ctx, AccrualServiceKey, accrualService)
+			ctx = context.WithValue(ctx, BalanceServiceKey, balanceService)
+
+			// Передаем управление следующему обработчику с обновленным контекстом
+			next.ServeHTTP(w, r.WithContext(ctx))
+		})
+	}
+}
+
+// GetServiceFromContext - извлекает сервис из контекста по заданному ключу.
+// В случае ошибки возвращает HTTP 500 и сообщение об ошибке.
+func GetServiceFromContext[Service interface{}](w http.ResponseWriter, r *http.Request, serviceKey key) *Service {
+	// Извлекаем сервис из контекста по ключу
+	foundService, ok := r.Context().Value(serviceKey).(Service)
+
+	if !ok {
+		// Если сервис не найден, возвращаем ошибку
+		http.Error(w, fmt.Sprintf("Сервис не найден в контексте по ключу %v", serviceKey), http.StatusInternalServerError)
+		return nil
+	}
+
+	return &foundService
+}
```

Comment:
👍 за использование дженериков. Советую использовать ключевое слово any вместо interface{}, так как оно какраз было введено для лучшей читаемости, как алиас.
Так же, если возвращать не указатель на `*Service`, а значение по-усполчанию можно избавиться от лишнего разименовывания там, где используется возврашаемое значение, те вместо `(*balanceService).GetUserBalance(r.Context(), user.ID)` -> `balanceService.GetUserBalance(r.Context(), user.ID)`.
Так же рекомендую возвращать ошибку в вместо вызова `http.Error(...)` и явно ее обрабатывать в вызывающем коде.

```suggestion
func GetServiceFromContext[Service any](w http.ResponseWriter, r *http.Request, serviceKey key) (Service, error) {
	// Извлекаем сервис из контекста по ключу
	var service Service
	if service, ok := r.Context().Value(serviceKey).(Service); !ok {
		// Если сервис не найден, возвращаем ошибку
		return service, fmt.Errorf("сервис не найден в контексте по ключу %", serviceKey)
	}

	return service, nil
}
```

Reply:
Пока не смог изменить ругаеться autotest

Reply:
Странно, мне удалось запустить такой вариант. Рекомендую разобраться с тем, что же все-таки не работает.